File: src/engines/default/DefaultPainter.ts
import { DefaultCanvasState } from "./DefaultCanvasState";
import { PainterDirection } from "../../PainterDirection";

export interface DefaultPainter extends WebAssembly.Instance {
    exports: {
        init(width: number, height: number): void;
        next(state: DefaultCanvasState, x: number, y: number): void;
        getLastActionColor(): number;
        getLastActionDirection(): PainterDirection;
    }
}

export function objIsDefaultPainter(instance: WebAssembly.Instance): instance is DefaultPainter {
    return true &&
        typeof instance.exports === "object" &&
        typeof instance.exports.init === "function" &&
        typeof instance.exports.next === "function" &&
        typeof instance.exports.getLastActionColor === "function" &&
        typeof instance.exports.getLastActionDirection === "function";
}

File: templates/default-painter.c
#include <stdio.h>
#include <emscripten.h>

ex

int g = 233;

EMSCRIPTEN_KEEPALIVE void init(int wivdth, int height) {
    g++;
}

EMSCRIPTEN_KEEPALIVE void next() {
}

EMSCRIPTEN_KEEPALIVE void getLastActionColor() {
}

EMSCRIPTEN_KEEPALIVE void getLastActionDirection() {
}